<a href="https://colab.research.google.com/github/pedrocostanunes/pedrocostanunes/blob/main/Data_Cleaning_OLIST.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Introdução

---
**Pedidos entregues:**

Qual a porcentagem de pedidos entregues dentro do prazo?

Qual é o tempo médio de entrega dos pedidos em cada estado?

Quais categorias de produto possuem menor tempo médio de entrega?

---
**Influência de preço e pagamento:**

O quanto o pagamento influencia no tempo de entrega?

Existe alguma correlação entre o tempo de entrega e o valor do frete?

Existe alguma correlação entre o tempo de entrega e o preço dos produtos?

Qual é o impacto do pagamento à vista ou parcelado no tempo de entrega

---
**Influência de timing do pedido:**

O dia da semana em que o pedido é realizado afeta o tempo de entrega?

O horário em que o pedido é realizado afeta o tempo de entrega?

---
**Influência das características do produto**

A categoria do produto influencia no tempo de entrega?

O peso dos produtos influencia no tempo de entrega?

## Importanto bibliotecas e carregando dados

Importanto bibliotecas

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from IPython.display import display, HTML

Carregando os dados

In [ ]:
clientes = pd.read_csv('/content/olist_customers_dataset.csv', sep =',')
pedidos = pd.read_csv('/content/olist_orders_dataset.csv', sep =',')
itens_pedidos = pd.read_csv('/content/olist_order_items_dataset.csv', sep =',')
pagamentos = pd.read_csv('/content/olist_order_payments_dataset.csv', sep =',')
produtos = pd.read_csv('/content/olist_products_dataset.csv', sep =',')
vendedores = pd.read_csv('/content/olist_sellers_dataset.csv', sep =',')
geolocalizacao = pd.read_csv('/content/olist_geolocation_dataset.csv', sep =',')

## Qualidade dos dados

#### Explorando com o método .info()

In [ ]:
clientes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [ ]:
pedidos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [ ]:
itens_pedidos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [ ]:
pagamentos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [ ]:
produtos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB


In [ ]:
vendedores.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB


In [ ]:
geolocalizacao.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB


#### Heatmap com resumo dos dados

In [ ]:
# Criando varável para armazenar os datasets e atribuindo títulos
datasets = [clientes, pedidos, itens_pedidos, pagamentos, produtos, vendedores, geolocalizacao]
titulos = ['clientes', 'pedidos', 'itens_pedidos', 'pagamentos', 'produtos', 'vendedores', 'geolocalizacao']

# Criando um dataframe para gerar o heatmap
resumo_dados = pd.DataFrame({},)
resumo_dados['datasets']= titulos
resumo_dados['linhas']= [data.shape[0] for data in datasets]
resumo_dados['colunas']= [data.shape[1] for data in datasets]
resumo_dados['duplicatas']= [data.duplicated().sum() for data in datasets]
resumo_dados['nulos']= [data.isnull().sum().sum() for data in datasets]

#gerando o heatmap
resumo_dados.style.background_gradient(cmap='RdPu').set_properties(**{'width': '120px', 'height': '20px'})

## Limpeza, transformação e preparação 

> Etapas da limpeza e transformação de dados

Com as análises anteriores pudemos identificar algumas limpezas e ajustes que podem ser feitos:

1. Identificar e tratar dados faltantes em cada DataFrame
2. Identificar e remover duplicatas
3. Unir todos os DFs em um único DataFrame para agilizar as análises
4. Transformar tipos de dados para os mais adequados
5. Traduzir colunas e campos para facilitar as leituras das análise
6. Criar novas colunas necessárias para auxiliar nas análises

#### 1. Identificar e tratar dados faltantes

Com a análise anterior, vimos que os DataFrames "pedidos" e "produtos" possuem dados nulos. Abaixo veremos a quantidade destes dados nulos e se eles são passiveis de remoção.

**1.1 Pedidos**

In [ ]:
pedidos.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

Neste DataFrame encontramos **4.908 campos nulos** em **3 colunas**.

Tendo em mente que nossa análise será feita com base em **pedidos entregues**,
vamos relacionar os dados nulos com a quantidade destes que tem seu
**'order_status' como 'delivered'**. Assim será possível decidir se podemos 
ou não remover os campos nulos do DF.

In [ ]:
pedidos_aprovados_nulos = pedidos.loc[pedidos['order_approved_at'].isnull()]
pedidos_aprovados_nulos.order_status.value_counts()

canceled     141
delivered     14
created        5
Name: order_status, dtype: int64

In [ ]:
pedidos_transportadora_nulos = pedidos.loc[pedidos['order_delivered_carrier_date'].isnull()]
pedidos_transportadora_nulos.order_status.value_counts()

unavailable    609
canceled       550
invoiced       314
processing     301
created          5
approved         2
delivered        2
Name: order_status, dtype: int64

In [ ]:
pedidos_entregues_nulos = pedidos.loc[pedidos['order_delivered_customer_date'].isnull()]
pedidos_entregues_nulos.order_status.value_counts()

shipped        1107
canceled        619
unavailable     609
invoiced        314
processing      301
delivered         8
created           5
approved          2
Name: order_status, dtype: int64

Das **4.908 linhas** com valores nulos neste DF, apenas **24 linhas** são relacionadas ao status de entrega **'delivered'**

Logo, vamos considerar que todas essas linhas com valores nulos podem ser removidas sem gerar impacto relevante em nossas análises. 

In [ ]:
pedidos = pedidos.dropna()
pedidos.isnull().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     0
order_delivered_customer_date    0
order_estimated_delivery_date    0
dtype: int64

1.2 Produtos

In [ ]:
produtos.isnull().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

O DF 'produtos' possui mais de 30.000 linhas. Encontramos apenas 610 destas com campos nulos em 4 colunas. Decidimos que também vamos remover estas linhas pois a baixa quantidade não afetará a análise.

In [ ]:
produtos = produtos.dropna()
produtos.isnull().sum()

product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64

#### 2. Identificar e remover duplicatas

Vimos no heatmap que apenas o dataframe **'geolocalizacao'** possui dados duplicados. 

Vamos simplismente removê-los.

In [ ]:
geolocalizacao.duplicated().sum()

261831

In [ ]:
geolocalizacao.drop_duplicates(inplace=True)
geolocalizacao.duplicated().sum()

0

#### 3. Unir os dados em um único dataframe

Não vamos incluir a 'geolocalizacao' neste merge. Existem colunas de zipcode tanto de clientes, quanto de vendedores e usar apenas uma delas para a junção pode acarretar na não inclusão de registros importantes no momento do merge.

Quando for necessário, vamos fazer a junção de 'geolocalizacao' para análises específicas.

In [ ]:
dados_olist= pd.merge(clientes, pedidos, on="customer_id")
dados_olist= dados_olist.merge(itens_pedidos, on="order_id")
dados_olist= dados_olist.merge(pagamentos, on="order_id")
dados_olist= dados_olist.merge(produtos, on="product_id")
dados_olist= dados_olist.merge(vendedores, on='seller_id')
dados_olist.shape

(113390, 33)

In [ ]:
dados_olist

customer_id                customer_unique_id  \
0       06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1       8912fc0c3bbf1e2fbf35819e21706718  9eae34bbd3a474ec5d07949ca7de67c0   
2       8912fc0c3bbf1e2fbf35819e21706718  9eae34bbd3a474ec5d07949ca7de67c0   
3       f0ac8e5a239118859b1734e1087cbb1f  3c799d181c34d51f6d44bbbc563024db   
4       6bc8d08963a135220ed6c6d098831f84  23397e992b09769faf5e66f9e171a241   
...                                  ...                               ...   
113385  59168d27c5da6a50e3d1f03d482e80b2  393c2ace23a135d0bcc7e70167448e16   
113386  eda10d7eb51a675d000956fa1aabc7dd  5c22dbb9d8b5385953f2f1e6d9d115a7   
113387  b90b857a277b080d8078332b3636ab58  7f5f43a9394964ac51bc5d1e0204ead4   
113388  1ed0c832c2dd99570a59260e71768bdf  82d46759af0369aad49084bacf85a6c3   
113389  1ed0c832c2dd99570a59260e71768bdf  82d46759af0369aad49084bacf85a6c3   

        customer_zip_code_prefix    customer_city customer_state  \
0                          14409           franca             SP   
1                          68030         santarem             PA   
2                          68030         santarem             PA   
3                          92480  nova santa rita             RS   
4                          25931             mage             RJ   
...                          ...              ...            ...   
113385                     47800        barreiras             BA   
113386                     39920         jordania             MG   
113387                     76400           uruacu             GO   
113388                     37610      bom repouso             MG   
113389                     37610      bom repouso             MG   

                                order_id order_status  \
0       00e7ee1b050b8499577073aeb2a297a1    delivered   
1       c1d2b34febe9cd269e378117d6681172    delivered   
2       c1d2b34febe9cd269e378117d6681172    delivered   
3       b1a5d5365d330d10485e0203d54ab9e8    delivered   
4       2e604b3614664aa66867856dba7e61b7    delivered   
...                                  ...          ...   
113385  c382d84d35f4d874b1f1c372f4bde27c    delivered   
113386  b494f624c631105b24fe497277085909    delivered   
113387  d76fe905f4588ad69ab3b36527bb267d    delivered   
113388  51c6d2f460589fa7b65f2da51e860206    delivered   
113389  51c6d2f460589fa7b65f2da51e860206    delivered   

       order_purchase_timestamp    order_approved_at  \
0           2017-05-16 15:05:35  2017-05-16 15:22:12   
1           2017-11-09 00:50:13  2017-11-10 00:47:48   
2           2017-11-09 00:50:13  2017-11-10 00:47:48   
3           2017-05-07 20:11:26  2017-05-08 22:22:56   
4           2018-02-03 19:45:40  2018-02-04 22:29:19   
...                         ...                  ...   
113385      2018-07-20 09:42:40  2018-07-20 15:05:22   
113386      2018-08-01 18:58:49  2018-08-02 18:50:15   
113387      2018-08-18 20:59:58  2018-08-18 21:10:13   
113388      2017-11-14 12:04:09  2017-11-14 12:15:25   
113389      2017-11-14 12:04:09  2017-11-14 12:15:25   

       order_delivered_carrier_date  ... product_name_lenght  \
0               2017-05-23 10:47:57  ...                41.0   
1               2017-11-22 01:43:37  ...                41.0   
2               2017-11-22 01:43:37  ...                41.0   
3               2017-05-19 20:16:31  ...                41.0   
4               2018-02-19 18:21:47  ...                41.0   
...                             ...  ...                 ...   
113385          2018-07-23 11:53:00  ...                49.0   
113386          2018-08-03 15:39:00  ...                51.0   
113387          2018-08-20 16:49:00  ...                55.0   
113388          2017-11-27 20:44:47  ...                33.0   
113389          2017-11-27 20:44:47  ...                33.0   

       product_description_lenght  product_photos_qty product_weight_g  \
0                          1141.0                 1.0           8683.0   


#### 4. Transformando tipos de dados para os mais adequados

In [ ]:
dados_olist.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 113390 entries, 0 to 113389
Data columns (total 33 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   customer_id                    113390 non-null  object 
 1   customer_unique_id             113390 non-null  object 
 2   customer_zip_code_prefix       113390 non-null  int64  
 3   customer_city                  113390 non-null  object 
 4   customer_state                 113390 non-null  object 
 5   order_id                       113390 non-null  object 
 6   order_status                   113390 non-null  object 
 7   order_purchase_timestamp       113390 non-null  object 
 8   order_approved_at              113390 non-null  object 
 9   order_delivered_carrier_date   113390 non-null  object 
 10  order_delivered_customer_date  113390 non-null  object 
 11  order_estimated_delivery_date  113390 non-null  object 
 12  order_item_id                 

**4.1 Colunas com datas**

Existem diversas colunas que possuem registros em timestamp. Vamos converter todas estas para **datetime**.

In [ ]:
colunas_datetime= ['order_purchase_timestamp', 'order_approved_at','order_delivered_carrier_date','order_delivered_customer_date',
               'order_estimated_delivery_date', 'shipping_limit_date']
dados_olist[colunas_datetime]=dados_olist[colunas_datetime].apply(pd.to_datetime)

**4.2 Colunas de zipcode**

As colunas de zipcode estão identificadas como int. Mas como são códigos que não serão usados para operações matemáticas, vamos convertê-las para string.

In [ ]:
dados_olist['customer_zip_code_prefix'] = dados_olist['customer_zip_code_prefix'].astype(str)
dados_olist['seller_zip_code_prefix'] = dados_olist['seller_zip_code_prefix'].astype(str)

**4.3 Colunas de ID**

A coluna '**order_item_id**' está como int64 e vamos converter para string.

In [ ]:
dados_olist['order_item_id'] = dados_olist['order_item_id'].astype(str)

**4.4 Colunas para Int**

Neste dataframe, as colunas '**product_name_lenght**', '**product_description_lenght**' e '**product_photos_qty**' apresentam valores inteiros e não precisam estar em formato float. Vamos transformar em int.   

In [ ]:
colunas_int = ['product_name_lenght', 'product_description_lenght', 'product_photos_qty']
dados_olist[colunas_int]=dados_olist[colunas_int].astype('int64')


**4.5 Verificando os tipos de dados**

In [ ]:
dados_olist.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 113390 entries, 0 to 113389
Data columns (total 33 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   customer_id                    113390 non-null  object        
 1   customer_unique_id             113390 non-null  object        
 2   customer_zip_code_prefix       113390 non-null  object        
 3   customer_city                  113390 non-null  object        
 4   customer_state                 113390 non-null  object        
 5   order_id                       113390 non-null  object        
 6   order_status                   113390 non-null  object        
 7   order_purchase_timestamp       113390 non-null  datetime64[ns]
 8   order_approved_at              113390 non-null  datetime64[ns]
 9   order_delivered_carrier_date   113390 non-null  datetime64[ns]
 10  order_delivered_customer_date  113390 non-null  datetime64[ns]
 11  

#### 5. Traduzindo colunas e registros

##### 5.1. Traduzindo as colunas

In [ ]:
dados_olist.columns

Index(['customer_id', 'customer_unique_id', 'customer_zip_code_prefix',
       'customer_city', 'customer_state', 'order_id', 'order_status',
       'order_purchase_timestamp', 'order_approved_at',
       'order_delivered_carrier_date', 'order_delivered_customer_date',
       'order_estimated_delivery_date', 'order_item_id', 'product_id',
       'seller_id', 'shipping_limit_date', 'price', 'freight_value',
       'payment_sequential', 'payment_type', 'payment_installments',
       'payment_value', 'product_category_name', 'product_name_lenght',
       'product_description_lenght', 'product_photos_qty', 'product_weight_g',
       'product_length_cm', 'product_height_cm', 'product_width_cm',
       'seller_zip_code_prefix', 'seller_city', 'seller_state'],
      dtype='object')

In [ ]:
colunas_traduzidas = {
    'customer_id':'cliente_id',
    'customer_unique_id':'cliente_id_unico', 
    'customer_zip_code_prefix':'cliente_cep_prefixo',
    'customer_city':'cliente_cidade', 
    'customer_state':'cliente_estado', 
    'order_id':'pedido_id', 
    'customer_id':'cliente_id',
    'order_status':'pedido_status', 
    'order_purchase_timestamp':'pedido_compra_datahora',
    'order_approved_at':'pedido_aprovado_datahora', 
    'order_delivered_carrier_date':'pedido_postagem_datahora',
    'order_delivered_customer_date':'pedido_entrega_datahora', 
    'order_estimated_delivery_date':'pedido_entrega_estimativa', 
    'order_item_id':'item_pedido_id', 
    'product_id':'produto_id', 
    'seller_id':'vendedor_id',
    'shipping_limit_date':'data_entrega_limite', 
    'price':'preco', 
    'freight_value':'valor_frete',
    'payment_sequential':'pagamento_sequencia', 
    'payment_type':'pagamento_tipo',
    'payment_installments':'pagamento_parcelas', 
    'payment_value':'pagamento_valor', 
    'product_category_name':'produto_categoria', 
    'product_name_lenght':'produto_nome_tamanho',
    'product_description_lenght':'produto_descricao_tamanho', 
    'product_photos_qty':'produto_fotos_qtd', 
    'product_weight_g':'produto_peso_g',
    'product_length_cm':'produto_comprimento_cm', 
    'product_height_cm':'produto_altura_cm', 
    'product_width_cm':'produto_largura_cm',
    'seller_zip_code_prefix':'vendedor_cep_prefixo',
    'seller_city':'vendedor_cidade', 
    'seller_state':'vendedor_estado'  
}

dados_olist.rename(columns=colunas_traduzidas, inplace=True)
dados_olist.head(3)

cliente_id                  cliente_id_unico  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  8912fc0c3bbf1e2fbf35819e21706718  9eae34bbd3a474ec5d07949ca7de67c0   
2  8912fc0c3bbf1e2fbf35819e21706718  9eae34bbd3a474ec5d07949ca7de67c0   

  cliente_cep_prefixo cliente_cidade cliente_estado  \
0               14409         franca             SP   
1               68030       santarem             PA   
2               68030       santarem             PA   

                          pedido_id pedido_status pedido_compra_datahora  \
0  00e7ee1b050b8499577073aeb2a297a1     delivered    2017-05-16 15:05:35   
1  c1d2b34febe9cd269e378117d6681172     delivered    2017-11-09 00:50:13   
2  c1d2b34febe9cd269e378117d6681172     delivered    2017-11-09 00:50:13   

  pedido_aprovado_datahora pedido_postagem_datahora  ... produto_nome_tamanho  \
0      2017-05-16 15:22:12      2017-05-23 10:47:57  ...                   41   
1      2017-11-10 00:47:48      2017-11-22 01:43:37  ...                   41   
2      2017-11-10 00:47:48      2017-11-22 01:43:37  ...                   41   

  produto_descricao_tamanho produto_fotos_qtd produto_peso_g  \
0                      1141                 1         8683.0   
1                      1141                 1         8683.0   
2                      1141                 1         8683.0   

  produto_comprimento_cm produto_altura_cm  produto_largura_cm  \
0                   54.0              64.0                31.0   
1                   54.0              64.0                31.0   
2                   54.0              64.0                31.0   

   vendedor_cep_prefixo  vendedor_cidade vendedor_estado  
0                  8577  itaquaquecetuba              SP  
1                  8577  itaquaquecetuba              SP  
2                  8577  itaquaquecetuba              SP  

[3 rows x 33 columns]

In [ ]:
dados_olist.columns

Index(['cliente_id', 'cliente_id_unico', 'cliente_cep_prefixo',
       'cliente_cidade', 'cliente_estado', 'pedido_id', 'pedido_status',
       'pedido_compra_datahora', 'pedido_aprovado_datahora',
       'pedido_postagem_datahora', 'pedido_entrega_datahora',
       'pedido_entrega_estimativa', 'item_pedido_id', 'produto_id',
       'vendedor_id', 'data_entrega_limite', 'preco', 'valor_frete',
       'pagamento_sequencia', 'pagamento_tipo', 'pagamento_parcelas',
       'pagamento_valor', 'produto_categoria', 'produto_nome_tamanho',
       'produto_descricao_tamanho', 'produto_fotos_qtd', 'produto_peso_g',
       'produto_comprimento_cm', 'produto_altura_cm', 'produto_largura_cm',
       'vendedor_cep_prefixo', 'vendedor_cidade', 'vendedor_estado'],
      dtype='object')

##### 5.2. Traduzindo valores de variáveis categóricas

pedido_status

In [ ]:
dados_olist['pedido_status'].unique()

array(['delivered', 'canceled'], dtype=object)

In [ ]:
traducao_pedido_status = {
    'delivered':'entregue', 
    'canceled':'cancelado'
 }

dados_olist['pedido_status'] = dados_olist['pedido_status'].map(traducao_pedido_status)
dados_olist['pedido_status'].unique()

array(['entregue', 'cancelado'], dtype=object)

pagamento_tipo

In [ ]:
dados_olist['pagamento_tipo'].unique()

array(['credit_card', 'boleto', 'voucher', 'debit_card'], dtype=object)

In [ ]:
traducao_pagamento_tipo = {'credit_card':'credito', 
                          'boleto':'boleto',
                          'voucher':'voucher',
                          'debit_card':'debito', 
                          'not_defined':'nao_informado'
 }

dados_olist['pagamento_tipo'] = dados_olist['pagamento_tipo'].map(traducao_pagamento_tipo)
dados_olist['pagamento_tipo'].unique()

array(['credito', 'boleto', 'voucher', 'debito'], dtype=object)

##### 5.3. [Extra] Padronizando os nomes de cidades

Vamos colocar as cidades com letra maiúscula em suas iniciais para facilitar a leitura e padronizar este tipo de dado. Temos duas colunas no dataframe para atualizar: 'cliente_cidade' e 'vendedor_cidade'

Obs.: Como algumas cidades possuem as palavras como 'de', 'do', 'da', vamos usar uma função para garantir que estas permaneçam minúsculas.

In [ ]:
dados_olist['cliente_cidade'] = dados_olist['cliente_cidade'].apply(lambda x: ' '.join(
    word.capitalize() if word not in ['de', 'do', 'da', 'dos', 'das']
                      else word
                      for word in x.split()
))

In [ ]:
dados_olist['vendedor_cidade'] = dados_olist['vendedor_cidade'].apply(lambda x: ' '.join(
    word.capitalize() if word not in ['de', 'do', 'da', 'dos', 'das'] 
                      else word
                      for word in x.split()
))

Verificando o output das funções acima:

In [ ]:
dados_olist['cliente_cidade'].unique()

array(['Franca', 'Santarem', 'Nova Santa Rita', ..., 'Cha Grande',
       'Bacaxa', 'Tangara'], dtype=object)

In [ ]:
dados_olist['vendedor_cidade'].unique()

array(['Itaquaquecetuba', 'Itajai', 'Ibitinga', 'Pirituba', 'Sao Paulo',
       'Rio de Janeiro', 'Curitiba', 'Piracicaba', 'Itatiba', 'Jacutinga',
       'Guaruja', 'Cariacica', 'Rio do Sul', 'Praia Grande',
       'Porto Alegre', 'Sao Jose do Rio Preto', 'Ribeirao Preto', 'Maua',
       'Toledo', 'Santo Andre', 'Limeira', 'Brasilia', 'Blumenau',
       'Campo Limpo Paulista', 'Sorocaba', 'Uberlandia', 'Cascavel',
       'Recife', 'Conchal', 'Vicente de Carvalho', 'Barueri', 'Campinas',
       'Betim', 'Sao Luis', 'Vassouras', 'Guariba', 'Pradopolis',
       'Sarandi', 'Maringa', 'Belo Horizonte', 'Sao Bernardo do Campo',
       'Indaiatuba', 'Mogi Mirim', 'Sumare', 'Pocos de Caldas',
       'Varzea Paulista', 'Foz do Iguacu', 'Sao Jose dos Campos',
       'Ampere', 'Sao Sebastiao', 'Salto', 'Pinhais', 'Fortaleza',
       'Japira', 'Tres Coracoes', 'Divinopolis', 'Rolandia', 'Claudio',
       'Braganca Paulista', 'Borda da Mata', 'Osasco', 'Poa', 'Guarulhos',
       'Francisco Beltrao

In [ ]:
"""No código acima encontramos uma 'cidade' que está registrada como e-mail 'Vendas@creditparts.com.br'.
   Em uma rápida pesquisa deste email combinado com o prefixo do CEP, encontramos que a empresa é,
   na verdade, de Maringá-PR. Então faremos essa rápida substituição para limpar este dado."""
   
dados_olist['vendedor_cidade'].replace('Vendas@creditparts.com.br', 'Maringá', inplace=True)

In [ ]:
dados_olist.vendedor_cidade.unique()

array(['Itaquaquecetuba', 'Itajai', 'Ibitinga', 'Pirituba', 'Sao Paulo',
       'Rio de Janeiro', 'Curitiba', 'Piracicaba', 'Itatiba', 'Jacutinga',
       'Guaruja', 'Cariacica', 'Rio do Sul', 'Praia Grande',
       'Porto Alegre', 'Sao Jose do Rio Preto', 'Ribeirao Preto', 'Maua',
       'Toledo', 'Santo Andre', 'Limeira', 'Brasilia', 'Blumenau',
       'Campo Limpo Paulista', 'Sorocaba', 'Uberlandia', 'Cascavel',
       'Recife', 'Conchal', 'Vicente de Carvalho', 'Barueri', 'Campinas',
       'Betim', 'Sao Luis', 'Vassouras', 'Guariba', 'Pradopolis',
       'Sarandi', 'Maringa', 'Belo Horizonte', 'Sao Bernardo do Campo',
       'Indaiatuba', 'Mogi Mirim', 'Sumare', 'Pocos de Caldas',
       'Varzea Paulista', 'Foz do Iguacu', 'Sao Jose dos Campos',
       'Ampere', 'Sao Sebastiao', 'Salto', 'Pinhais', 'Fortaleza',
       'Japira', 'Tres Coracoes', 'Divinopolis', 'Rolandia', 'Claudio',
       'Braganca Paulista', 'Borda da Mata', 'Osasco', 'Poa', 'Guarulhos',
       'Francisco Beltrao

In [ ]:
dados_olist.head(20)

cliente_id                  cliente_id_unico  \
0   06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1   8912fc0c3bbf1e2fbf35819e21706718  9eae34bbd3a474ec5d07949ca7de67c0   
2   8912fc0c3bbf1e2fbf35819e21706718  9eae34bbd3a474ec5d07949ca7de67c0   
3   f0ac8e5a239118859b1734e1087cbb1f  3c799d181c34d51f6d44bbbc563024db   
4   6bc8d08963a135220ed6c6d098831f84  23397e992b09769faf5e66f9e171a241   
5   fd3a0b1bd209f0e7d420c9c3d1127613  567ab47ca4deb92d46dbf54dce07d0a7   
6   fbd40c083aa8cddebb5265b2ba6aaf2e  f40ab89b622248b7ca125af4b486b887   
7   10558ef4afea173bfb5e2cbe3d5b0bb5  749943913a9851a39c9baf51877fbab6   
8   1c37c0f0cd1d88d46d9fc9494762abbd  432ecfa8b7b7ad2663c7abed0dc83c51   
9   20a452f528d487411fd7d3ebda1d0f20  31318a0597cd9d50ce4cfd03c80fe780   
10  20a452f528d487411fd7d3ebda1d0f20  31318a0597cd9d50ce4cfd03c80fe780   
11  62f766c82849282806bf2f62ca86432e  83aca63b2193ea126b287f3ae61528d9   
12  62f766c82849282806bf2f62ca86432e  83aca63b2193ea126b287f3ae61528d9   
13  62f766c82849282806bf2f62ca86432e  83aca63b2193ea126b287f3ae61528d9   
14  62f766c82849282806bf2f62ca86432e  83aca63b2193ea126b287f3ae61528d9   
15  62f766c82849282806bf2f62ca86432e  83aca63b2193ea126b287f3ae61528d9   
16  62f766c82849282806bf2f62ca86432e  83aca63b2193ea126b287f3ae61528d9   
17  d3286ca43aa21423c6045b2806ca8e56  acd02fc85cf6339f896c161af2b6abcd   
18  c49bec5f74f0413130abdaed8e7621d0  67d6f91d256bdf128fc960a166f413d5   
19  0058ebe2dc136d918dd001968cfa5903  9e8993ba2665307b71bd1641c2e1cb3d   

   cliente_cep_prefixo         cliente_cidade cliente_estado  \
0                14409                 Franca             SP   
1                68030               Santarem             PA   
2                68030               Santarem             PA   
3                92480        Nova Santa Rita             RS   
4                25931                   Mage             RJ   
5                88460               Angelina             SC   
6                32341               Contagem             MG   
7                78134          Varzea Grande             MT   
8                31270         Belo Horizonte             MG   
9                37540  Santa Rita do Sapucai             MG   
10               37540  Santa Rita do Sapucai             MG   
11               89115                 Gaspar             SC   
12               89115                 Gaspar             SC   
13               89115                 Gaspar             SC   
14               89115                 Gaspar             SC   
15               89115                 Gaspar             SC   
16               89115                 Gaspar             SC   
17               96083                Pelotas             RS   
18               25640             Petropolis             RJ   
19               39495            Montalvania             MG   

                           pedido_id pedido_status pedido_compra_datahora  \
0   00e7ee1b050b8499577073aeb2a297a1      entregue    2017-05-16 15:05:35   
1   c1d2b34febe9cd269e378117d6681172      entregue    2017-11-09 00:50:13   
2   c1d2b34febe9cd269e378117d6681172      entregue    2017-11-09 00:50:13   
3   b1a5d5365d330d10485e0203d54ab9e8      entregue    2017-05-07 20:11:26   
4   2e604b3614664aa66867856dba7e61b7      entregue    2018-02-03 19:45:40   
5   574fe1739f65af76badd0999db300b4f      entregue    2017-03-23 15:10:17   
6   e0b26f14d2bcc710bb02f77a4628763b      entregue    2017-05-16 10:00:49   
7   eaae5bd20fb15d85aa673d9b7c0e8ca5      entregue    2017-03-18 23:04:36   
8   c4e2bd2043fbd75b325a47adfabf0d77      entregue    2018-03-15 07:45:10   
9   1c7fe02ac4c7be50c59afb295cf85b89      entregue    2018-01-26 13:22:09   
10  1c7fe02ac4c7be50c59afb295cf85b89      entregue    2018-01-26 13:22:09   
11  5ee63c8c9ae2e5c38791b35932cf4c69      entregue    2017-05-09 11:30:18   
12  5ee63c8c9ae2e5c38791b35932cf4c69      entregue    2017-05-09 11:30:18   
13  5ee63c8c9ae2e5c38791b35932cf4c69      entreg

#### 6. Criando colunas de "tempo_entrega", "tempo_aprovado" e "dia_da_compra" para auxiliar nas análises

##### 6.1 "tempo_entrega"

In [ ]:
# Calcular a diferença entre as datas e converter para dias
dados_olist['tempo_entrega'] = (dados_olist['pedido_entrega_datahora'] - dados_olist['pedido_compra_datahora'])
dados_olist['tempo_entrega'] = dados_olist['tempo_entrega'].dt.days

Conferindo se existem valores negativos na coluna criada

In [ ]:
count = dados_olist[dados_olist['tempo_entrega'] < 0]['tempo_entrega'].count()
print(f"O número de linhas onde 'tempo_entrega' é menor que zero é: {count}")

O número de linhas onde 'tempo_entrega' é menor que zero é: 0


##### 6.2 'tempo_aprovacao'

In [ ]:
# O tempo de aprovação será calculado em horas e não em dias
dados_olist['tempo_aprovacao'] = (dados_olist['pedido_aprovado_datahora'] - dados_olist['pedido_compra_datahora'])
dados_olist['tempo_aprovacao'] = dados_olist['tempo_aprovacao'].dt.seconds/3600

Conferindo se existem valores negativos na nova coluna

In [ ]:
count = dados_olist[dados_olist['tempo_aprovacao'] < 0]['tempo_aprovacao'].count()
print(f"O número de linhas onde 'tempo_aprovacao' é menor que zero é: {count}")

O número de linhas onde 'tempo_aprovacao' é menor que zero é: 0


##### 6.3 'dia_da_compra'

In [ ]:
# adiciona uma nova coluna com o nome do dia da semana de cada pedido
dados_olist['dia_da_compra'] = dados_olist['pedido_compra_datahora'].dt.day_name()

In [ ]:
dados_olist.head()

cliente_id                  cliente_id_unico  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  8912fc0c3bbf1e2fbf35819e21706718  9eae34bbd3a474ec5d07949ca7de67c0   
2  8912fc0c3bbf1e2fbf35819e21706718  9eae34bbd3a474ec5d07949ca7de67c0   
3  f0ac8e5a239118859b1734e1087cbb1f  3c799d181c34d51f6d44bbbc563024db   
4  6bc8d08963a135220ed6c6d098831f84  23397e992b09769faf5e66f9e171a241   

  cliente_cep_prefixo   cliente_cidade cliente_estado  \
0               14409           Franca             SP   
1               68030         Santarem             PA   
2               68030         Santarem             PA   
3               92480  Nova Santa Rita             RS   
4               25931             Mage             RJ   

                          pedido_id pedido_status pedido_compra_datahora  \
0  00e7ee1b050b8499577073aeb2a297a1      entregue    2017-05-16 15:05:35   
1  c1d2b34febe9cd269e378117d6681172      entregue    2017-11-09 00:50:13   
2  c1d2b34febe9cd269e378117d6681172      entregue    2017-11-09 00:50:13   
3  b1a5d5365d330d10485e0203d54ab9e8      entregue    2017-05-07 20:11:26   
4  2e604b3614664aa66867856dba7e61b7      entregue    2018-02-03 19:45:40   

  pedido_aprovado_datahora pedido_postagem_datahora  ... produto_peso_g  \
0      2017-05-16 15:22:12      2017-05-23 10:47:57  ...         8683.0   
1      2017-11-10 00:47:48      2017-11-22 01:43:37  ...         8683.0   
2      2017-11-10 00:47:48      2017-11-22 01:43:37  ...         8683.0   
3      2017-05-08 22:22:56      2017-05-19 20:16:31  ...         8683.0   
4      2018-02-04 22:29:19      2018-02-19 18:21:47  ...         8683.0   

  produto_comprimento_cm produto_altura_cm produto_largura_cm  \
0                   54.0              64.0               31.0   
1                   54.0              64.0               31.0   
2                   54.0              64.0               31.0   
3                   54.0              64.0               31.0   
4                   54.0              64.0               31.0   

  vendedor_cep_prefixo  vendedor_cidade  vendedor_estado  tempo_entrega  \
0                 8577  Itaquaquecetuba               SP              8   
1                 8577  Itaquaquecetuba               SP             18   
2                 8577  Itaquaquecetuba               SP             18   
3                 8577  Itaquaquecetuba               SP             18   
4                 8577  Itaquaquecetuba               SP             25   

   tempo_aprovacao dia_da_compra  
0         0.276944       Tuesday  
1        23.959722      Thursday  
2        23.959722      Thursday  
3         2.191667        Sunday  
4         2.727500      Saturday  

[5 rows x 36 columns]

## Conclusão

FALTA:



1.   Escrever a indrodução
2.   Escrever a conclusão
3. Descobrir como subir isso pro GItHub
